### PR TITLE
Feat/hw fixes on x

### PR DESCRIPTION
**Pull request by ByteZhang1024**

### CODE REVIEW

**Comment by sidmorizon** [comment]
@codex review

**Comment by sidmorizon** [comment]
@codex security review

File: apps/desktop/app/process/BlePair.ts
Comment (sidmorizon) [inline apps/desktop/app/process/BlePair.ts:1]: **Cancelling BLE pairing shows a confusing hardware failure instead of a cancellation**

Reported at `apps/desktop/app/process/BlePair.ts:156-158` (outside the diff of this PR, so this is a file-level comment)

Severity: `non-severe`

The user-cancelled reason recorded for the pairing attempt is immediately replaced by the helper's own generic failure text (`lastError = event.message` at `apps/desktop/app/process/BlePair.ts:157`) before the attempt finishes, so a deliberate cancel is reported as an unexplained pairing failure.
Impact: A user who presses Cancel on the pairing-code dialog (or cancels from the app) sees a scary low-level Bluetooth failure message rather than a clean "cancelled" result.

<details>
<summary>How the cancel reason is overwritten before the promise rejects</summary>

When the decision callback receives `cancel` it sets `lastError = PAIR_CANCELLED_REASON` (`apps/desktop/app/process/BlePair.ts:109-111`), whose whole purpose is to match the SDK's `/connect cancelled/i` test so it maps to `BlePairingCancelled` (see comment at `apps/desktop/app/process/BlePair.ts:23-24`).

The helper then declines the ceremony, `PairAsync` returns a non-`Paired` status, and `run_pair` returns `Err(format!("pairing failed with status {status:?}"))` (`apps/desktop/native-modules/onekey-ble-pair/src/main.rs:787`). `main` turns that into `{"type":"error","message":"pairing failed with status DevicePairingResultStatus(N)"}` on stdout, which the stdout parser assigns to `lastError`, clobbering the cancel reason. The `exit` handler (`apps/desktop/app/process/BlePair.ts:175-190`) then rejects with the clobbered message.

A `cancelRequested` flag that makes the cancel reason win (or that skips overwriting `lastError` once a cancel has been sent) restores the intended mapping.

</details>

**Comment by originalix** [comment]
## Automated code review found blocking issues

@ByteZhang1024

Reviewed commit `cd5f3f110d17`. This report always reflects the latest reviewed changes.

### Review summary

This PR changes two hardware-wallet trust paths. On Windows desktop, Electron main and a per-attempt WinRT helper now keep numeric-comparison pairing pending until an explicit confirmation or cancellation; in Kaspa background signing, previous transaction IDs are resolved through the search and block endpoints so complete reference transactions can be streamed to the device. The refTx mapping is now isolated and tested, with local diagnostics and a blind-signing fallback, but two edge paths still bypass the intended safety behavior.

**What needs attention:** Confirm that every pairing termination, including timeout, reaches the WinRT refusal path and that lossless uint64 strings remain eligible for Kaspa refTx verification.

### Issues to address

- **P2 · Pairing timeout kills the helper before it can decline**
  - **Impact:** If the user leaves the numeric-comparison dialog unanswered for 60 seconds, the parent immediately terminates the helper. The helper's own refusal backstop is 120 seconds, so it cannot complete the WinRT deferral without accepting; Windows therefore cannot send the SMP pairing failure that this change relies on, and the device can observe the same dead-peer cancellation the PR is intended to fix.
  - **Suggested action:** On the parent timeout, send `cancel` through stdin and allow a short grace period for the helper to complete the deferral; reserve forced termination for a later backstop and align the two timeout values.
- **P2 · Exact uint64 strings unnecessarily disable refTx verification**
  - **Impact:** The block response contract permits exact decimal strings, but `readRefTxUint64` rejects every value above `Number.MAX_SAFE_INTEGER` regardless of whether it arrived as a rounded number or a lossless string. The hardware keyring catches that error and removes all `refTxs`, silently degrading the transaction to blind signing even though the device could have verified the exact uint64 value.
  - **Suggested action:** Apply the safe-integer restriction only to numeric inputs, and validate string inputs as non-negative integers within the uint64 range; add tests for exact large strings, negative values, and overflow.

### Validation gaps

- `git diff --check origin/x...HEAD` completed without whitespace errors.
- The targeted Jest test could not start because this worktree has no Yarn `node_modules` state file.
- The Windows helper could not be compiled locally because the Rust toolchain is absent, and the WinRT confirmation/cancellation flow was not exercised on Windows hardware.
- The bundled PE files have the expected x64 and ARM64 architectures and contain the new decision-path strings, but reproducible source-to-binary equivalence was not established.
- Existing GitHub review and comment coverage could not be retrieved because GitHub network access failed, so inline comments require a human duplicate check.

**Comment by sidmorizon** [comment]
@cursoragent review

File: packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  }
+  const uint64 = (
+    value: number | string | null | undefined,
+    field: string,
+    nullMeansZero: boolean,
+  ) => readRefTxUint64({ value, field, txId, networkId, nullMeansZero });
+
+  return {
+    txId,
+    version: tx.version,
```
Comment (sidmorizon) [inline packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts:83]: # 🟠 P1 tx.version 为 0 时会被省略

## 问题

`/blocks/{blockId}?includeTransactions=true` 的响应来自 kaspad gRPC 的 protobuf → JSON 转换，proto3 默认值字段整体不落键（响应用的正是 `previousOutpoint` / `subnetworkId` / `verboseData` 这套 camelCase 命名，符合 `MessageToDict` 的行为）。这也正是本 PR 对 `sequence`、`lockTime`、`gas`、`scriptPublicKey.version`、`previousOutpoint.index`、`payload` 都做了缺失兜底的原因。

但 `IKaspaBlockTransaction.version` 被声明成必填 `number`，这里也直接透传 `version: tx.version`，是唯一没有兜底的数值字段。

## 影响

v0 交易的 `version` 恰好等于 `0`，响应里不会带这个键，于是 `tx.version === undefined`。紧接着 `Vault.collectRefTxsByApi` 的门禁 `refTxs.some((tx) => tx.version !== 0)` 恒为真并抛出 `unsupported non-v0 prev tx`，`KeyringHardware` 捕获后置 `refTxs = undefined`。

结果是本 PR 想启用的设备端输入校验，对 `100%` 的 v0 前序交易都会静默退化成盲签，仅在本地日志留下一条 `refTxFetchFailed`——而 v0 正是这条链路唯一支持的版本。

## 建议

把 `version` 也按「缺失即 0」读取，并同步放宽类型：

```ts
// clientRestApi.ts
version?: number | string | null;

// refTxUtils.ts
version: Number(uint64(tx.version, 'version', true)),
```

同时建议补一条「version 缺失时按 v0 处理并通过门禁」的用例，现有测试的 `blockTx()` 始终显式写了 `version: 0`，覆盖不到这个真实响应形态。

[Claude session ↗](https://claude.ai/code/session_01MkpLWw9pkfEVxChbKRG6Pt)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  field: string;
+  txId: string;
+  networkId: string | undefined;
+  nullMeansZero: boolean;
+}): string {
+  if (value === null || value === undefined) {
+    if (!nullMeansZero) {
+      throw new OneKeyLocalError(`kaspa refTx: ${field} missing for ${txId}`);
+    }
+    defaultLogger.transaction.send.refTxFieldDefaulted({
```
Comment (sidmorizon) [inline packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts:32]: # 🟡 P2 零值兜底日志会淹没自身信号

## 问题

零值字段缺失在这条链路上是常态而非异常：几乎每个 input 的 `sequence`、每笔 tx 的 `lockTime` 与 `gas`、每个 output 的 `scriptPublicKey.version` 实际值都是 `0`，因而都会走进 `nullMeansZero` 分支，各写一条 `refTxFieldDefaulted` 本地日志。

## 影响

一笔 `90` 输入的交易要拉最多 `90` 笔前序交易，每笔按 1~2 个 input、2 个 output 估算约 `6` 条，单次硬件签名就会写入约 `540` 条本地日志。`send.ts` 里这条日志的注释说明它的用途是「某个本该有值的字段被读成 0 时可追溯」，但它在正常路径上高频触发，真正异常的那一条会被完全淹没，同时挤占本地日志容量。

## 建议

改成按交易聚合一条，把被兜底的字段名与次数一起带出：

```ts
// buildKaspaRefTx 内收集，末尾一次性上报
if (defaultedFields.length) {
  defaultLogger.transaction.send.refTxFieldDefaulted({
    network: networkId,
    txId,
    field: defaultedFields.join(','),
  });
}
```

或者仅对「零值不属于预期」的字段记录日志，`sequence` / `lockTime` / `gas` / `scriptVersion` 这类常态零值直接静默返回 `'0'`。

[Claude session ↗](https://claude.ai/code/session_01MkpLWw9pkfEVxChbKRG6Pt)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by sidmorizon — COMMENTED** [review]
Review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts (hunk 1 above)
Comment (sidmorizon) [reply packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts:83]: 在 `91769c7ec5` 复核：数据路径未变，`refTxUtils.ts:69` 仍是 `version: tx.version`，类型仍是必填 `number`，`Vault.ts:831` 的 `refTxs.some((tx) => tx.version !== 0)` 门禁也原样保留，所以本条继续成立。

这一轮新加的注释反而把矛盾摆到了同一个文件里：`refTxUtils.ts:9-11` 现在写着「api.kaspa.org does send those as 0, so it is only a guard」。这两种说法只能成立一个——

- 若上游确实显式下发 `0`，那么 `sequence` / `lockTime` / `gas` / `scriptPublicKey.version` 的 `nullMeansZero` 分支与 `refTxFieldDefaulted` 日志就是永不触发的死代码，`clientRestApi.ts` 里把这些字段声明成可选/可空也没有必要；
- 若上游会按 proto3 省略默认值（这正是那些兜底存在的理由），那 `tx.version` 就是唯一没兜底的数值字段，v0 前序交易会稳定命中 `unsupported non-v0 prev tx`，本 PR 的设备端输入校验对 100% 的 v0 交易静默退化成盲签。

建议直接用一次真实响应确认，然后把结论落到代码而不是注释上：要么给 `version` 补同样的「缺失即 0」读取并放宽类型，要么移除其余字段的兜底与日志。测试里的 `blockTx()` 目前总是显式写 `version: 0`，覆盖不到缺键这一真实形态，补一条用例即可锁住行为。

[Claude session ↗](https://claude.ai/code/session_01MkpLWw9pkfEVxChbKRG6Pt)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by originalix — COMMENTED** [review]
@ByteZhang1024

## Automated code review found blocking issues

Reviewed commit `91769c7ec5d7`.

- **P1 · Stale pairing dialogs can answer a later ceremony**
- **P2 · Exact uint64 strings unnecessarily disable refTx verification**

[View the full review report](https://github.com/OneKeyHQ/app-monorepo/pull/12911#issuecomment-5317689480)

**Review by originalix — COMMENTED** [review]
@ByteZhang1024

## Automated code review found blocking issues

Reviewed commit `f5264130d0be`.

- **P1 · Stale pairing dialogs can answer a later ceremony**
- **P2 · Exact uint64 strings unnecessarily disable refTx verification**

[View the full review report](https://github.com/OneKeyHQ/app-monorepo/pull/12911#issuecomment-5317689480)

File: apps/desktop/app/process/trezorBlePairing.ts [hunk 3]
```diff
@@ -122,16 +118,30 @@ export function createTrezorBlePairingIpcMain(
...
+        const decision = response === 1 ? 'cancel' : 'confirm';
+        // Compare against the helper's `sinceAccept` to tell an OS timeout apart
+        // from "we outran the user".
         logger.info(
-          `[TrezorBLE] pin dialog dismissed after ${Date.now() - shownAt}ms`,
+          `[TrezorBLE] pin dialog answered '${decision}' after ${
+            Date.now() - shownAt
+          }ms`,
         );
+        const delivered = decideActivePairing(decision);
```
Comment (sidmorizon) [inline apps/desktop/app/process/trezorBlePairing.ts]: # 🟠 P1 rebase 之后整层 BLE 配对代码已无调用方

## 问题

本次推送是把分支 rebase 到当前 `x`（base 由 `40c4d68796` 变为 `f61081f6ac`）。PR 自身 11 个文件的内容 diff 与上一个 head `f5264130d0` 逐字节相同，说明 rebase 过程没有任何冲突需要解决。

问题正出在「没有冲突」上。旧 base 的 `apps/desktop/app/app.ts:1692` 原本是：

```ts
initTrezorBleSupport(browserWindow.webContents, {
  ipcMain: createTrezorBlePairingIpcMain(
    trezorBleSenderGatedIpcMain,
    browserWindow,
  ),
```

而当前 `x`（`923d2f36b8` "chore: sync hotfix/v6.5.2 to x"）已经把这层包装摘掉，新 base 的 `app.ts:1667-1668` 直接传裸的 `trezorBleSenderGatedIpcMain`。本 PR 从未修改过 `app.ts`，所以 rebase 静默通过、没有任何提示。全仓库检索 `createTrezorBlePairingIpcMain`，现在只剩 `trezorBlePairing.ts:76` 的定义本身，**零调用方**。

## 影响

`showPin`（包括本 PR 新增的 cancel 按钮和这一行的 `decideActivePairing`）、`ensurePaired` 的串行化、新增的 `cancelPairing` 分支，以及 `BlePair.ts` 的 `ensureDevicePaired` / `decideActivePairing` / helper 拉起，全部不可达。Windows 上 connect 之前不再执行 WinRT OS 配对，本 PR 要解决的「noble 无法发起 bonding、GATT 报 Unreachable」会原样复现；scan 的 id→address 缓存和 raw-advertisement 取证也一并失效。两个约 `212KB` 的 `onekey-ble-pair.exe` 仍然随包发布，但永远不会被执行。

死代码照常通过 tsc 与 oxlint，主进程接线又没有测试覆盖，所以本次 CI 全绿并不能说明这条链路还活着。

## 建议

先确认 `@onekeyfe/hwk-trezor-connector-electron-ble` 从 `1.1.34-alpha.2` 升到 `1.2.0-alpha.183` 之后，Windows OS 配对是否已由 SDK 自己承担。

若否，在 `app.ts` 恢复包装：

```ts
initTrezorBleSupport(browserWindow.webContents, {
  ipcMain: createTrezorBlePairingIpcMain(
    trezorBleSenderGatedIpcMain,
    browserWindow,
  ),
  // ...
```

若是，则本 PR 的 desktop BLE 部分（含两个 `.exe`）应整体删除，而不是留在树里无人调用。无论哪一种，都不应停在「改了一个没人调用的模块」这个中间状态。

[Claude session ↗](https://claude.ai/code/session_01MkpLWw9pkfEVxChbKRG6Pt)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: apps/desktop/app/process/BlePair.ts
Comment (ByteZhang1024) [reply apps/desktop/app/process/BlePair.ts:1]: 已修,f8ec5f8cc0

File: apps/desktop/native-modules/onekey-ble-pair/src/main.rs [hunk 4]
```diff
@@ -83,6 +83,99 @@ fn init_log_file() {
...
+static DECISION: AtomicU8 = AtomicU8::new(DECISION_PENDING);
+
+/// Poll interval while the delegate waits for the parent's decision.
+#[cfg(windows)]
+const DECISION_POLL_MS: u64 = 25;
+
+/// Backstop so a parent that never answers cannot wedge the helper; the device
+/// gives up on its own pairing window long before this.
+#[cfg(windows)]
+const DECISION_TIMEOUT_MS: u64 = 120_000;
```
[the remark this answers is not in the document]
Comment (ByteZhang1024) [reply apps/desktop/native-modules/onekey-ble-pair/src/main.rs:105]: 已修,f8ec5f8cc0

File: packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts (hunk 1 above)
Comment (ByteZhang1024) [reply packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts:83]: 零值字段一个都没被省略,tx.version === undefined 不成立,"100% v0 交易退化成盲签"没有复现。

File: packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts (hunk 2 above)
Comment (ByteZhang1024) [reply packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts:32]: 基于同一个前提,同样不成立。零值字段(sequence / lockTime / gas /
scriptPublicKey.version)在真实响应里都带值,nullMeansZero 分支根本进不去,
所以"一次签名 540 条日志"不会发生。

**Review by originalix — COMMENTED** [review]
@ByteZhang1024

## Automated code review found blocking issues

Reviewed commit `92a1babe42df`.

- **P1 · Stale pairing dialogs can answer a later ceremony**

[View the full review report](https://github.com/OneKeyHQ/app-monorepo/pull/12911#issuecomment-5317689480)

**Review by originalix — COMMENTED** [review]
@ByteZhang1024

## Automated code review found blocking issues

Reviewed commit `7c92807fc39a`.

- **P1 · Stale pairing dialogs can answer a later ceremony**

[View the full review report](https://github.com/OneKeyHQ/app-monorepo/pull/12911#issuecomment-5317689480)

**Review by originalix — COMMENTED** [review]
@ByteZhang1024

## Automated code review found blocking issues

Reviewed commit `2b95f287d192`.

- **P1 · Stale pairing dialogs can answer a later ceremony**
- **P1 · The timeout can kill a helper after valid confirmation**

[View the full review report](https://github.com/OneKeyHQ/app-monorepo/pull/12911#issuecomment-5317689480)

**Review by originalix — COMMENTED** [review]
@ByteZhang1024

## Automated code review found blocking issues

Reviewed commit `959f5762305e`.

- **P1 · Stale pairing dialogs can answer a later ceremony**
- **P1 · The timeout can kill a helper after valid confirmation**

[View the full review report](https://github.com/OneKeyHQ/app-monorepo/pull/12911#issuecomment-5317689480)

File: apps/desktop/app/process/trezorBlePairing.ts (hunk 3 above)
Comment (ByteZhang1024) [reply apps/desktop/app/process/trezorBlePairing.ts]: 修复 959f576230

File: packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    nullMeansZero: boolean,
+  ) => readRefTxUint64({ value, field, txId, networkId, nullMeansZero });
+
+  return {
+    txId,
+    version: tx.version,
+    inputs: (tx.inputs ?? []).map((input) => ({
+      prevTxId: input.previousOutpoint.transactionId,
+      outputIndex: Number(input.previousOutpoint.index ?? 0),
+      sequenceNumber: uint64(input.sequence, 'input.sequence', true),
```
[the remark this answers is not in the document]
Comment (ByteZhang1024) [reply packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts:87]: 修复 7c92807fc3

File: packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      throw new OneKeyLocalError(`kaspa refTx: ${field} missing for ${txId}`);
+    }
+    defaultLogger.transaction.send.refTxFieldDefaulted({
+      network: networkId,
+      txId,
+      field,
+    });
+    return '0';
+  }
+  const parsed = new BigNumber(String(value));
```
[the remark this answers is not in the document]
Comment (ByteZhang1024) [reply packages/kit-bg/src/vaults/impls/kaspa/refTxUtils.ts:39]: 修复 92a1babe42

**Review by originalix — COMMENTED** [review]
@ByteZhang1024

## Automated code review found blocking issues

Reviewed commit `28e41633ad70`.

- **P1 · Confirmed BLE pairing has no completion watchdog**

[View the full review report](https://github.com/OneKeyHQ/app-monorepo/pull/12911#issuecomment-5317689480)

**Review by zhaono1 — APPROVED** [review]

**Review by originalix — APPROVED** [review]
